run JS in chunks on page load

johnduffell opened the pull request:
Our JS was taking a long time to run on mobile causing a lot of jank
![image](https://cloud.githubusercontent.com/assets/7304387/8854960/8235879e-315a-11e5-839f-cd2987a051c3.png)
![image](https://cloud.githubusercontent.com/assets/7304387/8854968/911e30b2-315a-11e5-93d6-854dc6284a98.png)
so I've split it to run each block on a new animation frame:
![image](https://cloud.githubusercontent.com/assets/7304387/8855077/1a8eed50-315b-11e5-9b3f-b865a1ae77be.png)
![image](https://cloud.githubusercontent.com/assets/7304387/8855106/3dadeaa2-315b-11e5-91fd-56b215b701f2.png)
This means the JS takes a lot longer to run overall, but it's chunked into small pieces.  Some of the chunks take way way less than 16ms to run, so could probably be blocked together a bit better.  Some of them could do with being further split down.. Omniture is a bad offender.
@sndrs what do you think?

OliverJAsh commented on static/src/javascripts/bootstraps/facia.js:
```diff
@@ -114,17 +118,19 @@ define([
+                    ['f-weather', modules.showWeather],
+                    ['f-live-blog-updates', modules.showLiveblogUpdates],
+                    ['f-slideshow', modules.startSlideshow]
+                ]));
```
You shouldn't need to wrap these calls in `robust.catchErrorsAndLog`, because the functions are already executed asynchronously using `requestAnimationFrame`. Any errors will be observed by `window.onerror` which reports to Sentry.

johnduffell replied:
good point thanks!  although I do still want to add the next animation frame even if there's an error.  Maybe if I do that first it would cover that possibility though - I'll do that.

OliverJAsh commented:
This will be an interesting experiment! :smile: How are we going to measure the difference in perceived performance?

I wonder if there's a noticeable difference in time it takes for the page to become fully interactive.

OliverJAsh commented on static/src/javascripts/projects/common/utils/background.js:
```diff
@@ -0,0 +1,21 @@
+                    restFunctions();
+                });
+            };
+        }, function () {}))();
```
Could you just do a `forEach`? You're not using the returned value?

OliverJAsh replied:
``` js
codeBlocks.forEach(function (fn) { rAF(fn); });
```

?

OliverJAsh replied:
Ah, of course you need to queue the animation frames, ignore me!

OliverJAsh commented on static/src/javascripts/projects/common/utils/background.js:
```diff
@@ -0,0 +1,21 @@
+            return function () {
+                requestAnimationFrame(function () {
+                    fn();
+                    restFunctions();
```
You can guarantee the next chunk will be executed by reversing these: queue the next animation frame, then execute the current chunk.

OliverJAsh commented:
> Some of the chunks take way way less than 16ms to run, so could probably be blocked together a bit better

Whereas some chunks will take longer than 16ms, presumably? Also depends on device.

Ideally each frame would execute a small amount of JS (<5ms), to ensure enough time is left for the user to interact with the page and any subsequent event bound JS to execute. The trade off is how long it takes for the page to become fully interactive. Perhaps we could defer initialisation of individual features until we detect a user is about to use a feature.

Is the goal here to make sure the site feels fast as soon as the user begins interacting with it (usually scrolling immediately on load)?

paperboyo commented:
> Is the goal here to make sure the site feels fast as soon as the user begins interacting with it (usually scrolling immediately on load)?

That would be massive! CGW was always the fastest to show up from the whole wide world. Still is. But slightly, over time, became this see-how-fast-I-show-up-just-do-not-try-to-touch-me-cause-I-aint-moving kind of thing. Personal experience, though, of a guy with the ancient phone.

johnduffell commented:
@OliverJAsh 
> Perhaps we could defer initialisation of individual features until we detect a user is about to use a feature.
That's an excellent point, at the moment it's all in whatever hard coded order we fancied.
> Is the goal here to make sure the site feels fast as soon as the user begins interacting with it (usually scrolling immediately on load)?
Absolutely, and this was even worse - we have a lot of initial jank from 3 full style calculations when you arrive on the page, but that's all in the first few frames.  However this next bit was tacked on the end once the JS came in, so it could really interrupt at the wrong moment especially on slow connections.
@paperboyo absolutely, well many people have ancient phones/slow connections and this navigation slowness is discouraging people from navigating, which is a bad thing when we want to encourage more pages per visit

piuccio commented:
I tried to implement this already in the past (I can dig the PR if you want), let me tell you what went wrong.

Some modules call `mediator.on('ready)` and wait for `common.js` to emit `ready`.

This is the code right now
```
// Inside common.js
// Inlining modules functions
mediator.on('ready', weather))
mediator.on('ready', somethingElse))

mediator.emit('ready')
```

If you make the module init async this is what happens

`
raf(() => mediator.on('ready', weather))
raf(() => mediator.on('ready', somethingElse))

mediator.emit('ready')
`

So the `ready` event is emitted before all module registered an interest.

Check that you don't have the same problem.

johnduffell commented:
@piuccio I've hopefully addressed those issues you mentioned, thanks.  I also added a switch so I can turn it off quickly when things go pear shaped ;)

piuccio commented:
@OliverJAsh was it useful to give a name to those functions? How does sentry look like in case of errors?

OliverJAsh commented:
> Absolutely, and this was even worse - we have a lot of initial jank from 3 full style calculations when you arrive on the page, but that's all in the first few frames. However this next bit was tacked on the end once the JS came in, so it could really interrupt at the wrong moment especially on slow connections.

Does this mean we're still going ahead with the PR?

OliverJAsh commented:
@piuccio You mean https://github.com/guardian/frontend/pull/9918/files#diff-3357de000cb815b9e493a795ad1b1560R24? `fn.name` is set for function declarations, but not function expressions—so it depends how the function is defined, AFAIK. I know some interpreters are trying to improve this to work for function expressions too.

piuccio commented:
@OliverJAsh 

before
```js
robust.catchErrorsAndLogAll([
       ['c-analytics', modules.loadAnalytics],
```

after
```js
background([
      modules.loadAnalytics,
```

Was `c-analytics` useful?

OliverJAsh commented:
Good point @piuccio.

johnduffell commented:
can I get a +1 or some suggestions from someone please?

The idea is to release this as an improvement, to break the 9am gridlock, but to make something else to avoid the rush hour altogether by considering the needs of each piece of code separately and running them at the right point.

[This long document](https://www.chromium.org/developers/design-documents/gpu-accelerated-compositing-in-chrome) is really useful to read although heavy going, and actually janky scroll isn't really a thing - webkit can scroll the viewport without waiting for the main thread as it all happens on the compositor.  That's not to say the compositor might be too slow or the main thread may slow down the compositor thread, but we only need to worry about the 16ms thing if we are actually producing an animation from the main thread, not just scrolling.  As it is we just have to worry about the 100ms time to respond to interactions, which does need the main thread.

@OliverJAsh @sndrs @piuccio

OliverJAsh commented:
> janky scroll isn't really a thing - webkit can scroll the viewport without waiting for the main thread as it all happens on the compositor. That's not to say the compositor might be too slow or the main thread may slow down the compositor thread, but we only need to worry about the 16ms thing if we are actually producing an animation from the main thread, not just scrolling

Good overview of rendering in Blink. One exception!

Janky scrolling is caused by:

* lazy loading containers (which we only do on mobile) (which will trigger a repaint of the entire document layer)
* naughty ads

It might be worth investigating the former if it is a bottleneck? The latter is a much bigger problem and something we should try and protect ourselves from, and/or monitor. I'm exploring doing so with Telemetry, although would appreciate the help from you guys on Reach if you think it's worthwhile. Smoothness is incredibly important to a news website IMO.

By "animation" do you mean anything that changes a style or manipulates the DOM (as these will cause a repaint)?

OliverJAsh commented:
Given your analysis, do we still need to load the JS in chunks?

piuccio commented:
considering that optimization is most useful for slow browser, this should at least be tested on a mobile / old browser. Ship it unless it makes things worse.

Btw the problem we're trying to solve here is scrolling during page load, right?

johnduffell commented:
@OliverJAsh  it would still be useful as with the changes we're not blocking the main thread any more.  Although the main thread isn't needed for scrolling, it is needed for clicking on stories.
@piuccio I've mainly been testing it on my nexus 5 (so should test on something slower)
the problem I *was* trying to solve was scrolling during load, but I couldn't detect any improvement during manual testing.  Eventually it turns out we didn't have that problem because webkit is smart.  However not blocking the main thread is a good idea anyway so I still think it's worth shipping.

OliverJAsh commented:
> it is needed for clicking on stories.

So have you noticed a big difference in perceived performance for tap/click interactions during load?

piuccio commented:
:+1: unless it requires some testing (it probably does :grin: )

OliverJAsh commented:
> So have you noticed a big difference in perceived performance for tap/click interactions during load?

Thinking about this more, I'm really interested to know if this has made a difference for these interactions. Beforehand, the main thread was blocked by JS execution, so there would surely be a big delay before interactions were handled. A good test would be: how fast does the browser navigate when you tap the top link when the page is still loading?
